src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs

```diff
@@ -0,0 +1,661 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
+using System.Text;
+
+namespace System.Security.Cryptography
+{
+    /// <summary>
+    ///   NIST SP 800-108 HMAC CTR Key-Based Key Derivation (KBKDF)
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     This implements NIST SP 800-108 HMAC in counter mode. The implemented KDF assumes the form of
+    ///     <c>PRF (KI, [i]2 || Label || 0x00 || Context || [L]2)</c> where <c>[i]2</c> and <c>[L]2</c> are encoded as
+    ///     unsigned 32-bit integers, big endian.
+    ///   </para>
+    ///   <para>
+    ///     All members of this class are thread safe. If the instance is disposed of while other threads are using
+    ///     the instance, those threads will either receive an <see cref="ObjectDisposedException" /> or produce a valid
+    ///     derived key.
+    ///   </para>
+    /// </remarks>
+    public sealed partial class SP800108HmacCounterKdf : IDisposable
+    {
+        private static readonly bool s_useCngKeyDerivation = IsWindows8OrGreater();
+
+#if !NET7_0_OR_GREATER && NET
+        [UnsupportedOSPlatformGuard("browser")]
+        private static bool ManagedSupported => !OperatingSystem.IsBrowser();
+#else
+        private static bool ManagedSupported => true;
+#endif
+
+        // The maximum amount of data that we can produce with the PRF is 0x1FFFFFFF.
+        // This is because of L[2]. From SP 800-108 r1:
+        // L – An integer specifying the requested length (in bits) of the derived keying material KOUT.
+        // As an unsigned 32-bit interger (see r), L needs to become L[2] by multiplying by 8 (bytes to bits).
+        // We can't encode more than 0x1FFFFFFF as bits without overflowing.
+        // Windows' BCryptKeyDerivation cannot fullfill a request larger than 0x1FFFFFFF, either.
+        private const int MaxPrfOutputSize = (int)(uint.MaxValue / 8);
+
+        private readonly SP800108HmacCounterKdfImplementationBase _implementation;
+
+        /// <summary>
+        ///   Initializes a new instance of <see cref="SP800108HmacCounterKdf" /> using key and HMAC algorithm.
```

Comment:
There's a bit of cave-man speech going on here.  I think the standard form I've seen is

```suggestion
        ///   Initializes a new instance of <see cref="SP800108HmacCounterKdf" /> using a specified key and HMAC algorithm.
```

(also applies to the next ctor)

Reply:
SP800-108 good. Cave man speech bad!